configfile: "config.yaml"

import pandas as pd
samples = pd.read_csv(config["samples"], index_col="row ID", sep="\t")	#load samples into table
#samples = pd.read_csv(config["samples"], index_col="accession")	#load samples into table

########### Bowtie index variables #########
reference_file = config["reference"]	# reference sequence file
index = config["index"]		# bowtie index

INDEX_EXT = ["1.bt2", "2.bt2", "3.bt2", "4.bt2", "rev.1.bt2", "rev.2.bt2"]
###############################################



# target rule
rule all:
	input:
		# Bowtie index -> is created if only the reference file exists
		expand("{index_name}.{index_ext}", index_name=index, index_ext=INDEX_EXT),
		
		# FastQC before trimming
    	expand("output/stats/fastqc_original/{sample}_1.html", sample=list(samples.index)),
    	expand("output/stats/fastqc_original/{sample}_2.html", sample=list(samples.index)),
    	
    	# FastQC after trimming
    	expand("output/stats/fastqc_trimmed/{sample}_1.html", sample=list(samples.index)),
    	expand("output/stats/fastqc_trimmed/{sample}_2.html", sample=list(samples.index)),
    	
    	# Mapping -> not needed to be listed
    	#expand("output/mapped/{sample}.bam", sample=list(samples.index)),
    	 
    	# Mapping stats
    	expand("output/mapped/{sample}.bam.idxstats", sample=list(samples.index)),
    	 
    	# MulitQC stats -> including FastQC & Qualimap
    	"output/stats/multiqc.html",
    	
		# Consensus Sequence -> not needed to be listed
		#expand("output/ivar/{sample}.fa", sample=list(samples.index)),
		#expand("output/ivar/{sample}.qual.txt", sample=list(samples.index)),
		
		# Multiple alignment
		#"output/augur/aligned.fasta",
		#"output/augur/tree_raw.nwk",
		"output/augur/phylogenetic_tree.png",
		"output/augur/entropies.txt",
		"output/augur/entropies.png"



include:	"rules/bowtie.smk"		# Creates bowtie-index, 4. runs mapping
include:	"rules/qc.smk"			# 1. FastQC before trim 2. Trim 3. FastQC after trim, 8. Qualimap for mappings, 9. MultiQC
include:	"rules/samtools.smk"	# 5. sort bam-files, 6. Samtools-index, 7. Samtools-mapping-stats

include:	"rules/ivar.smk"		# Genome assembly
include:	"rules/augur.smk"		# Multiple alignment, Phylo Tree, Entropies
